HSEARCH-5068 Simplify entity registration in the Standalone POJO mapper

yrodiere opened the pull request:
https://hibernate.atlassian.net/browse/HSEARCH-5068

Here you go. Best reviewed commit by commit, as it's huge (sorry) and some commits are really not worth your time (see below).

The first few commits are various fixes for problems I noticed while working on this issue; not completely relevant, but useful nonetheless.

Starting with "HSEARCH-5068 Make it possible to contribute entity/indexed metadata from multiple sources" we're getting in the "refactoring" realm: mostly moving stuff from `mapper-pojo-standalone`/`mapper-orm` to `pojo-base`. Not useful per se, but necessary for the actual changes. I wish it wasn't, I really do, and to be frank I'm ashamed it was, as it more or less highlights how stuff wasn't designed well enough to evolve easily. But here we are.
IMO this is not worth reviewing, as I told you @marko-bekhta . Feel free to have a look if you think it's necessary, but please be mindful of our time constraints (sorry -- I did as fast as I could).

With "HSEARCH-5068 Introduce AnnotatedTypeSource to easily configure annotated types" is a sort of a taster: this commit makes it more obvious to users that they need to tell us where to look for annotations, because soon it'll be possible to perform all the mapping using *only* annotations. See [our discussion on Zulip](https://hibernate.zulipchat.com/#narrow/stream/132092-hibernate-search-dev/topic/HSEARCH-5068).

The next commit "HSEARCH-5068 Move definition of entities to pojo-base and add `@SearchEntity`" is the real stuff, where the magic happens. This introduces a `@SearchEntity` annotation to replace the old `addEntity` stuff. Hopefully the code, plus the documentation added in this commit, make it all rather understandable.

The final commit "HSEARCH-5068 Move from addEntity to @SearchEntity/.searchEntity in integration tests" is only about getting rid of all references to the deprecated `addEntity` methods, so again, IMO not worth reviewing.

Now... let's hope CI agrees with me :D

yrodiere commented:
Looks like there are a couple things to fix to make JQAssistant happy, and I just noticed I forgot to update the migration guide (I only changed incubating APIs and a few SPIs, but still).

I'll add fix-up commits this morning to address that.

marko-bekhta commented:
JQAssistant always gets grumpy after classes are moved around 😃 🙈

yrodiere commented:
> I'll add fix-up commits this morning to address that.

Done :crossed_fingers:

marko-bekhta reviewed (approved):
nice 😃 !

I assume this works well with Quarkus, right? it seems there're a few tests failing in the docs, but otherwise I only had a few questions/comments 😃.

aaand one more question:

> to track progress on allowing the use of `@SearchEntity` in the Hibernate ORM integration to map non-ORM entities

I was wondering if `@SearchEntity` could still be useful on an `@Entity` to maybe redefine a loading binder? (I'm thinking about https://github.com/hibernate/hibernate-search/pull/3920)

yrodiere commented on mapper/pojo-standalone/src/main/java/org/hibernate/search/mapper/pojo/standalone/mapping/impl/StandalonePojoEntityTypeContributor.java:
```diff
@@ -22,13 +22,11 @@ class StandalonePojoEntityTypeContributor implements PojoTypeMetadataContributor
 	}
 
 	@Override
-	// Keeping the deprecated form in order to test that it works correctly (for Infinispan in particular)
```
If they didn't, that's on them. This was deprecated in 6.2...

yrodiere commented on documentation/src/main/asciidoc/public/reference/_mapping-configuration.adoc:
```diff
@@ -174,13 +174,38 @@ However, `AnnotationMappingConfigurationContext`
+
+Hibernate Search will call the `configure` method of this implementation on startup,
+and the configurer will be able to take advantage of a DSL to define
+the programmatic mapping, for example:
```
Right; done.

yrodiere commented on mapper/orm/src/main/java/org/hibernate/search/mapper/orm/mapping/impl/HibernateOrmEntityTypeMetadataContributor.java:
```diff
@@ -33,10 +33,9 @@ public void contributeAdditionalMetadata(PojoAdditionalMetadataCollectorTypeNode
-				new HibernateOrmPathDefinitionProvider( typeModel, persistentClass )
-		)
-				.entityIdPropertyName( identifierPropertyNameOptional.orElse( null ) );
+		var node = collector.markAsEntity();
```
Feel free to have a look, but keep in mind that, in documentation, an explicit type is actually useful: it helps understand what classes are involved.

So, personally, I would only do this in examples where the type of a variable is already very, very obvious.

yrodiere commented on mapper/pojo-base/src/main/java/org/hibernate/search/mapper/pojo/model/additionalmetadata/building/impl/PojoEntityTypeAdditionalMetadataBuilder.java:
```diff
@@ -12,41 +12,47 @@
-				entityName,
-				pathDefinitionProvider,
+				entityName != null ? entityName : typeModel.typeIdentifier().javaClass().getSimpleName(),
+				pathDefinitionProvider != null ? pathDefinitionProvider : new SimplePojoPathsDefinitionProvider(),
```
Alright, done

yrodiere commented on mapper/pojo-base/src/main/java/org/hibernate/search/mapper/pojo/mapping/definition/annotation/AnnotatedTypeSource.java:
```diff
@@ -0,0 +1,86 @@
+	 * @return A source that {@link AnnotationMappingConfigurationContext#add(Set) adds} the provided classes explicitly,
+	 * and by default also allows {@link AnnotationMappingConfigurationContext#discoverJandexIndexesFromAddedTypes(boolean) other classes from the same JAR to be discovered automatically}
+	 * if they are {@link AnnotationMappingConfigurationContext#discoverAnnotatedTypesFromRootMappingAnnotations(boolean) annotated with root mapping annotations}
+	 * such as {@link }.
```
Right, fixed

yrodiere commented on mapper/pojo-base/src/main/java/org/hibernate/search/mapper/pojo/mapping/definition/annotation/AnnotatedTypeSource.java:
```diff
@@ -0,0 +1,86 @@
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.search.mapper.pojo.mapping.definition.annotation;
```
In theory it could be used for ORM too. I was contemplating just moving all such config over from `AnnotationMappingConfigurationContext` to `AnnotatedTypeSource`, but yeah, we'll have another look when we deal with HSEARCH-5076.

In any case, I marked this class as `@Incubating`, like everything in the Standalone Pojo Mapper, so that we're free to move it around.

yrodiere commented on mapper/pojo-standalone/src/main/java/org/hibernate/search/mapper/pojo/standalone/bootstrap/impl/StandalonePojoIntegrationBooterImpl.java:
```diff
@@ -130,12 +138,19 @@ public StandalonePojoMapping boot() {
 		}
 
+		@Override
+		public BuilderImpl annotatedTypeSource(AnnotatedTypeSource source) {
```
It's mainly for integrators, e.g. Quarkus indeed. I think users can achieve the same effect with a mapping configurer, so I wouldn't bother.

yrodiere commented on mapper/pojo-standalone/src/main/java/org/hibernate/search/mapper/pojo/standalone/mapping/SearchMapping.java:
```diff
@@ -126,18 +127,22 @@ default <T> SearchScope<T> scope(Class<T> expectedSuperType, String entityName)
 	 */
-	static SearchMappingBuilder builder() {
-		return new SearchMappingBuilder();
+	static SearchMappingBuilder builder(AnnotatedTypeSource annotatedTypeSource) {
```
Removing `builder()` (no param) was the whole point :)

I want the API to make it very clear that Hibernate Search needs some input to be able to process annotations. Otherwise that's too easily missed.

Those who want only programmatic mapping will have pass `AnnotatedTypeSource.empty()` explicitly.

yrodiere commented on documentation/src/main/asciidoc/public/reference/_mapper-pojo-standalone.adoc:
```diff
@@ -34,21 +34,25 @@ at the moment.
+through <<mapping-classpath-scanning,classpath scanning>>,
+which should work well on simple (single-mapping) setups.
+
+For more complex setups, consider <<mapping-classpath-scanning-faster,tuning classpath scanning>>
+and/or using a <<mapping-configurer,mapping configurer>>
+to <<mapper-orm-mapping-configuration-scanning-dependencies,provide the list of annotated types explicitly>>.
```
Right... I wrote that before I introduced the `AnnotatedTypeSource` thing. I'll rework it.

yrodiere commented:
Thanks!

> I assume this works well with Quarkus, right? it seems there're a few tests failing in the docs, but otherwise I only had a few questions/comments 😃.

Unfortunately I didn't have time to test that :(

I know that I need this `@SearchEntity` annotation though, so let's release a CR1, and hopefully the changes required for Quarkus (if any) won't be large enough to require a CR2.

> > to track progress on allowing the use of `@SearchEntity` in the Hibernate ORM integration to map non-ORM entities
>
> I was wondering if `@SearchEntity` could still be useful on an @Entity to maybe redefine a loading binder? (I'm thinking about https://github.com/hibernate/hibernate-search/pull/3920)

It could, see HSEARCH-5080.

---

Otherwise I think I addressed your comments. I'll fix the docs tests, rebase, and let's hope we can merge then.
